Inline crashing compiler after cyclic reference detected.

``` scala
  inline def foo(n: Int) = foo(n)
```

```
tests/main.scala:7: error: cyclic reference involving method foo
  inline def foo(n: Int) = foo(n)
                           ^
exception occurred while typechecking foo.scala
exception occurred while compiling foo.scala
Exception in thread "main" dotty.tools.dotc.ast.Trees$UnAssignedTypeException: type of Ident(n) is not assigned
    at dotty.tools.dotc.ast.Trees$Tree.tpe(Trees.scala:93)
    at dotty.tools.dotc.ast.Trees$DenotingTree.denot(Trees.scala:226)
    at dotty.tools.dotc.ast.Trees$Tree.symbol(Trees.scala:144)
    at dotty.tools.dotc.typer.Inliner$addAccessors$2$.transform(Inliner.scala:143)
    at dotty.tools.dotc.ast.Trees$Instance$TreeMap$$anonfun$transform$2.apply(Trees.scala:1161)
    at dotty.tools.dotc.ast.Trees$Instance$TreeMap$$anonfun$transform$2.apply(Trees.scala:1161)
    at scala.collection.immutable.List.loop$1(List.scala:173)
    at scala.collection.immutable.List.mapConserve(List.scala:189)
    at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1161)
    at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1080)
    at dotty.tools.dotc.typer.Inliner$addAccessors$2$.transform(Inliner.scala:141)
    at dotty.tools.dotc.typer.Inliner$.dotty$tools$dotc$typer$Inliner$$makeInlineable(Inliner.scala:171)
    at dotty.tools.dotc.typer.Inliner$$anonfun$registerInlineInfo$1.apply(Inliner.scala:194)
    at dotty.tools.dotc.typer.Inliner$$anonfun$registerInlineInfo$1.apply(Inliner.scala:191)
    at dotty.tools.dotc.core.Annotations$LazyBodyAnnotation.tree(Annotations.scala:70)
    at dotty.tools.dotc.typer.Inliner$.bodyAndAccessors(Inliner.scala:207)
    at dotty.tools.dotc.typer.Inliner$.bodyToInline(Inliner.scala:216)
    at dotty.tools.dotc.typer.Typer.inlineExpansion(Typer.scala:1538)
    at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1516)
    at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1530)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:1248)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:1196)
    at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
    at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:1196)
    at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1432)
    at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1480)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1492)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1490)
    at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:154)
    at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
    at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1490)
    at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1514)
    at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1530)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:1348)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:1339)
    at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
    at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:1339)
    at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1470)
    at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1482)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1492)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1490)
    at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:154)
    at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
    at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1490)
    at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1542)
    at dotty.tools.dotc.typer.FrontEnd$$anonfun$typeCheck$1.apply$mcV$sp(FrontEnd.scala:50)
    at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:24)
    at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:48)
    at dotty.tools.dotc.typer.FrontEnd$$anonfun$runOn$3.apply(FrontEnd.scala:72)
    at dotty.tools.dotc.typer.FrontEnd$$anonfun$runOn$3.apply(FrontEnd.scala:72)
    at scala.collection.immutable.List.foreach(List.scala:381)
    at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:72)
    at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:71)
    at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:68)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
    at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply$mcV$sp(Run.scala:68)
    at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:62)
    at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:62)
    at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:76)
    at dotty.tools.dotc.Run.compileUnits(Run.scala:62)
    at dotty.tools.dotc.Run.compileSources(Run.scala:59)
    at dotty.tools.dotc.Run.compile(Run.scala:43)
    at dotty.tools.dotc.Driver.doCompile(Driver.scala:26)
    at dotty.tools.dotc.Driver.process(Driver.scala:124)
    at dotty.tools.dotc.Driver.process(Driver.scala:93)
    at dotty.tools.dotc.Driver.process(Driver.scala:105)
    at dotty.tools.dotc.Driver.main(Driver.scala:132)
    at dotty.tools.dotc.Main.main(Main.scala)
```

--- a/compiler/src/dotty/tools/dotc/ast/Trees.scala
+++ b/compiler/src/dotty/tools/dotc/ast/Trees.scala
@@ -1069,95 +1069,100 @@ object Trees {
         Template(tree: Tree)(constr, parents, self, body)
     }
 
+    /** Hook to indicate that a transform of some subtree should be skipped */
+    protected def skipTransform(tree: Tree)(implicit ctx: Context): Boolean = false
+
     abstract class TreeMap(val cpy: TreeCopier = inst.cpy) {
 
-      def transform(tree: Tree)(implicit ctx: Context): Tree = tree match {
-        case Ident(name) =>
-          tree
-        case Select(qualifier, name) =>
-          cpy.Select(tree)(transform(qualifier), name)
-        case This(qual) =>
-          tree
-        case Super(qual, mix) =>
-          cpy.Super(tree)(transform(qual), mix)
-        case Apply(fun, args) =>
-          cpy.Apply(tree)(transform(fun), transform(args))
-        case TypeApply(fun, args) =>
-          cpy.TypeApply(tree)(transform(fun), transform(args))
-        case Literal(const) =>
-          tree
-        case New(tpt) =>
-          cpy.New(tree)(transform(tpt))
-        case Typed(expr, tpt) =>
-          cpy.Typed(tree)(transform(expr), transform(tpt))
-        case NamedArg(name, arg) =>
-          cpy.NamedArg(tree)(name, transform(arg))
-        case Assign(lhs, rhs) =>
-          cpy.Assign(tree)(transform(lhs), transform(rhs))
-        case Block(stats, expr) =>
-          cpy.Block(tree)(transformStats(stats), transform(expr))
-        case If(cond, thenp, elsep) =>
-          cpy.If(tree)(transform(cond), transform(thenp), transform(elsep))
-        case Closure(env, meth, tpt) =>
-          cpy.Closure(tree)(transform(env), transform(meth), transform(tpt))
-        case Match(selector, cases) =>
-          cpy.Match(tree)(transform(selector), transformSub(cases))
-        case CaseDef(pat, guard, body) =>
-          cpy.CaseDef(tree)(transform(pat), transform(guard), transform(body))
-        case Return(expr, from) =>
-          cpy.Return(tree)(transform(expr), transformSub(from))
-        case Try(block, cases, finalizer) =>
-          cpy.Try(tree)(transform(block), transformSub(cases), transform(finalizer))
-        case SeqLiteral(elems, elemtpt) =>
-          cpy.SeqLiteral(tree)(transform(elems), transform(elemtpt))
-        case Inlined(call, bindings, expansion) =>
-          cpy.Inlined(tree)(call, transformSub(bindings), transform(expansion))
-        case TypeTree() =>
-          tree
-        case SingletonTypeTree(ref) =>
-          cpy.SingletonTypeTree(tree)(transform(ref))
-        case AndTypeTree(left, right) =>
-          cpy.AndTypeTree(tree)(transform(left), transform(right))
-        case OrTypeTree(left, right) =>
-          cpy.OrTypeTree(tree)(transform(left), transform(right))
-        case RefinedTypeTree(tpt, refinements) =>
-          cpy.RefinedTypeTree(tree)(transform(tpt), transformSub(refinements))
-        case AppliedTypeTree(tpt, args) =>
-          cpy.AppliedTypeTree(tree)(transform(tpt), transform(args))
-        case PolyTypeTree(tparams, body) =>
-          cpy.PolyTypeTree(tree)(transformSub(tparams), transform(body))
-        case ByNameTypeTree(result) =>
-          cpy.ByNameTypeTree(tree)(transform(result))
-        case TypeBoundsTree(lo, hi) =>
-          cpy.TypeBoundsTree(tree)(transform(lo), transform(hi))
-        case Bind(name, body) =>
-          cpy.Bind(tree)(name, transform(body))
-        case Alternative(trees) =>
-          cpy.Alternative(tree)(transform(trees))
-        case UnApply(fun, implicits, patterns) =>
-          cpy.UnApply(tree)(transform(fun), transform(implicits), transform(patterns))
-        case EmptyValDef =>
-          tree
-        case tree @ ValDef(name, tpt, _) =>
-          val tpt1 = transform(tpt)
-          val rhs1 = transform(tree.rhs)
-          cpy.ValDef(tree)(name, tpt1, rhs1)
-        case tree @ DefDef(name, tparams, vparamss, tpt, _) =>
-          cpy.DefDef(tree)(name, transformSub(tparams), vparamss mapConserve (transformSub(_)), transform(tpt), transform(tree.rhs))
-        case tree @ TypeDef(name, rhs) =>
-          cpy.TypeDef(tree)(name, transform(rhs))
-        case tree @ Template(constr, parents, self, _) =>
-          cpy.Template(tree)(transformSub(constr), transform(parents), transformSub(self), transformStats(tree.body))
-        case Import(expr, selectors) =>
-          cpy.Import(tree)(transform(expr), selectors)
-        case PackageDef(pid, stats) =>
-          cpy.PackageDef(tree)(transformSub(pid), transformStats(stats))
-        case Annotated(arg, annot) =>
-          cpy.Annotated(tree)(transform(arg), transform(annot))
-        case Thicket(trees) =>
-          val trees1 = transform(trees)
-          if (trees1 eq trees) tree else Thicket(trees1)
-      }
+      def transform(tree: Tree)(implicit ctx: Context): Tree =
+        if (skipTransform(tree)) tree
+        else tree match {
+          case Ident(name) =>
+            tree
+          case Select(qualifier, name) =>
+            cpy.Select(tree)(transform(qualifier), name)
+          case This(qual) =>
+            tree
+          case Super(qual, mix) =>
+            cpy.Super(tree)(transform(qual), mix)
+          case Apply(fun, args) =>
+            cpy.Apply(tree)(transform(fun), transform(args))
+          case TypeApply(fun, args) =>
+            cpy.TypeApply(tree)(transform(fun), transform(args))
+          case Literal(const) =>
+            tree
+          case New(tpt) =>
+            cpy.New(tree)(transform(tpt))
+          case Typed(expr, tpt) =>
+            cpy.Typed(tree)(transform(expr), transform(tpt))
+          case NamedArg(name, arg) =>
+            cpy.NamedArg(tree)(name, transform(arg))
+          case Assign(lhs, rhs) =>
+            cpy.Assign(tree)(transform(lhs), transform(rhs))
+          case Block(stats, expr) =>
+            cpy.Block(tree)(transformStats(stats), transform(expr))
+          case If(cond, thenp, elsep) =>
+            cpy.If(tree)(transform(cond), transform(thenp), transform(elsep))
+          case Closure(env, meth, tpt) =>
+            cpy.Closure(tree)(transform(env), transform(meth), transform(tpt))
+          case Match(selector, cases) =>
+            cpy.Match(tree)(transform(selector), transformSub(cases))
+          case CaseDef(pat, guard, body) =>
+            cpy.CaseDef(tree)(transform(pat), transform(guard), transform(body))
+          case Return(expr, from) =>
+            cpy.Return(tree)(transform(expr), transformSub(from))
+          case Try(block, cases, finalizer) =>
+            cpy.Try(tree)(transform(block), transformSub(cases), transform(finalizer))
+          case SeqLiteral(elems, elemtpt) =>
+            cpy.SeqLiteral(tree)(transform(elems), transform(elemtpt))
+          case Inlined(call, bindings, expansion) =>
+            cpy.Inlined(tree)(call, transformSub(bindings), transform(expansion))
+          case TypeTree() =>
+            tree
+          case SingletonTypeTree(ref) =>
+            cpy.SingletonTypeTree(tree)(transform(ref))
+          case AndTypeTree(left, right) =>
+            cpy.AndTypeTree(tree)(transform(left), transform(right))
+          case OrTypeTree(left, right) =>
+            cpy.OrTypeTree(tree)(transform(left), transform(right))
+          case RefinedTypeTree(tpt, refinements) =>
+            cpy.RefinedTypeTree(tree)(transform(tpt), transformSub(refinements))
+          case AppliedTypeTree(tpt, args) =>
+            cpy.AppliedTypeTree(tree)(transform(tpt), transform(args))
+          case PolyTypeTree(tparams, body) =>
+            cpy.PolyTypeTree(tree)(transformSub(tparams), transform(body))
+          case ByNameTypeTree(result) =>
+            cpy.ByNameTypeTree(tree)(transform(result))
+          case TypeBoundsTree(lo, hi) =>
+            cpy.TypeBoundsTree(tree)(transform(lo), transform(hi))
+          case Bind(name, body) =>
+            cpy.Bind(tree)(name, transform(body))
+          case Alternative(trees) =>
+            cpy.Alternative(tree)(transform(trees))
+          case UnApply(fun, implicits, patterns) =>
+            cpy.UnApply(tree)(transform(fun), transform(implicits), transform(patterns))
+          case EmptyValDef =>
+            tree
+          case tree @ ValDef(name, tpt, _) =>
+            val tpt1 = transform(tpt)
+            val rhs1 = transform(tree.rhs)
+            cpy.ValDef(tree)(name, tpt1, rhs1)
+          case tree @ DefDef(name, tparams, vparamss, tpt, _) =>
+            cpy.DefDef(tree)(name, transformSub(tparams), vparamss mapConserve (transformSub(_)), transform(tpt), transform(tree.rhs))
+          case tree @ TypeDef(name, rhs) =>
+            cpy.TypeDef(tree)(name, transform(rhs))
+          case tree @ Template(constr, parents, self, _) =>
+            cpy.Template(tree)(transformSub(constr), transform(parents), transformSub(self), transformStats(tree.body))
+          case Import(expr, selectors) =>
+            cpy.Import(tree)(transform(expr), selectors)
+          case PackageDef(pid, stats) =>
+            cpy.PackageDef(tree)(transformSub(pid), transformStats(stats))
+          case Annotated(arg, annot) =>
+            cpy.Annotated(tree)(transform(arg), transform(annot))
+          case Thicket(trees) =>
+            val trees1 = transform(trees)
+            if (trees1 eq trees) tree else Thicket(trees1)
+        }
 
       def transformStats(trees: List[Tree])(implicit ctx: Context): List[Tree] =
         transform(trees)

--- a/compiler/src/dotty/tools/dotc/ast/tpd.scala
+++ b/compiler/src/dotty/tools/dotc/ast/tpd.scala
@@ -573,6 +573,8 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
       Try(tree: Tree)(expr, cases, finalizer)
   }
 
+  override def skipTransform(tree: Tree)(implicit ctx: Context) = tree.tpe.isError
+
   implicit class TreeOps[ThisTree <: tpd.Tree](val tree: ThisTree) extends AnyVal {
 
     def isValue(implicit ctx: Context): Boolean =

--- a/tests/neg/i1568.scala
+++ b/tests/neg/i1568.scala
@@ -1,0 +1,3 @@
+object Test {
+  inline def foo(n: Int) = foo(n) // error: cyclic reference
+}

